### ISSUE TITLE
[NJT] Dropout with p=0 still drops values

### ISSUE THREAD

**Issue by imh** [pytorch/pytorch#151717]
### 🐛 Describe the bug

I'm trying to track down discrepancies where an NJT version of my model diverges in training while the padded version converges nicely.

I found a discrepancy in dropout between NJT and normal tensors that was interfering with reproducibility efforts: Setting dropout to 0 does not prevent dropout from happening with NJTs.

Here's as minimal of a repro as I've been able to make:

```python
from collections import Counter

import torch
import torch.nn.functional as F

# set up data and state
values = torch.randn((6598, 384), device='cuda', dtype=torch.float32)
offsets = torch.arange(129, device='cuda', dtype=torch.int64)
offsets[-1] = values.shape[0]
rng_state = torch.tensor([0, 0, 0, 0, 0, 0, 0, 0, 176, 0, 0, 0, 0, 0, 0, 0], dtype=torch.uint8)
torch.cuda.set_rng_state(rng_state)
# before and after dropout
before = torch.nested.nested_tensor_from_jagged(values, offsets)
after = F.dropout(before, p=0.0)
# expect them to be the same, but they aren't
print(f"{(before != after).sum()=}")  # 1 discrepancy
idx = (before != after)._values.nonzero().squeeze(0)  # at [189, 342], an insignificant location
print(f"{before._values[*idx]=}")  # 1.3 before
print(f"{after._values[*idx]=}")  #  0 after
```

It happens about 7% of calls with NJT and exactly 0% with normal tensors:

```python
# happens about 7% of the runs
discrepancies = Counter()
for i in range(10_000):
    after = F.dropout(before, p=0.0)
    num_mismatched = (before != after).sum().item()
    discrepancies[num_mismatched] += 1
print(f"{discrepancies=}")  # {0: 9275, 1: 707, 2: 18}
# but its limited to NJTs
discrepancies = Counter()
for i in range(10_000):
    after = F.dropout(before._values, p=0.0)
    num_mismatched = (before._values != after).sum().item()
    discrepancies[num_mismatched] += 1
print(f"{discrepancies=}")  # {0: 10000}
```

### Versions

Collecting environment information...
PyTorch version: 2.6.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.2 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.12.3 (main, Feb  4 2025, 14:48:35) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-57-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version: 565.57.01
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.4
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        39 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               8
On-line CPU(s) list:                  0-7
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz
CPU family:                           6
Model:                                94
Thread(s) per core:                   2
Core(s) per socket:                   4
Socket(s):                            1
Stepping:                             3
CPU(s) scaling MHz:                   96%
CPU max MHz:                          4200.0000
CPU min MHz:                          800.0000
BogoMIPS:                             7999.96
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault pti ssbd ibrs ibpb stibp tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp vnmi md_clear flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            128 KiB (4 instances)
L1i cache:                            128 KiB (4 instances)
L2 cache:                             1 MiB (4 instances)
L3 cache:                             8 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-7
Vulnerability Gather data sampling:   Vulnerable: No microcode
Vulnerability Itlb multihit:          KVM: Mitigation: VMX disabled
Vulnerability L1tf:                   Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:                    Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:               Mitigation; PTI
Vulnerability Mmio stale data:        Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; IBRS; IBPB conditional; STIBP conditional; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Mitigation; Microcode
Vulnerability Tsx async abort:        Mitigation; TSX disabled

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect

**Comment by soulitzer** [comment]
> I'd be more than happy to help fix it if someone could guide me a bit on the C++ side. I haven't ventured into that side of torch.

@imh Sounds great. This shouldn't be c++ related by the way, I'd start by looking at what happens in `torch/nested/_internal/ops.py`

**Comment by imh** [comment]
I had tracked it to at least [here](https://github.com/pytorch/pytorch/blob/5d316ce0d080d9393ae4ac501c724aac239841c2/torch/nested/_internal/ops.py#L859-L872) (i've commented the values around the function call):

```python
@register_jagged_func(
    torch.ops.aten.native_dropout.default, "self: jt, float: any, train: any?"
)
def native_dropout_default(func, *args, **kwargs):
    _, new_kwargs = normalize_function(  # type: ignore[misc]
        func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
    )

    inp = new_kwargs.pop("input")
    # new_kwargs: {'p': 0.0, 'train': True}
    # func: <OpOverload(op='aten.native_dropout', overload='default')>
    out1, out2 = func(inp._values, **new_kwargs)
    # (out1 != inp._values).sum(): 1
    # (~out2).sum(): 1
    return (
        NestedTensor(out1, **extract_kwargs(inp)),
        NestedTensor(out2, **extract_kwargs(inp)),
    )
```

The surprising dropping out happens in the call to `<OpOverload(op='aten.native_dropout', overload='default')>`, which *seems* like it should be benign since it's called with a normal tensor (`inp._values`), but I wasn't able to follow into that call (I figured it was C++ past that).

Any docs on what calling `<OpOverload(op='aten.native_dropout', overload='default')>` does? Suggestions of where to continue hunting this down?

**Comment by jbschlosser** [comment]
Huh super weird - thanks for calling this out! On the C++ side, native_dropout is defined here for CPU:
https://github.com/pytorch/pytorch/blob/21b0ef520d651ed67f6978ac37c8a8a4093819ee/aten/src/ATen/native/Dropout.cpp#L104-L125

and here for CUDA:
https://github.com/pytorch/pytorch/blob/21b0ef520d651ed67f6978ac37c8a8a4093819ee/aten/src/ATen/native/cuda/Dropout.cu#L419-L437
